### PR TITLE
Initial Implementation of Chat Functionality

### LINKED ISSUE #213: Initial Webchat API
The API should provide the following functionalities:

- (POST) Create a text message
- (POST) Uplead a picture
- (GET) Retrieve chat history
- (GET) Download a picture

The API should be placed at /api/v1/projects/{projectId}/tasks/{taskId}/chat
and for defects at /api/v1/projects/{projectId}/defects/{defectId}/chat

### PR DESCRIPTION
**Pull request by parhamrahmani**
###  Initial Implementation of Chat Functionality  - 

Closes #213
Related to #59 

---

#### **Overview**

This PR introduces chat functionality to the project, designed for tasks and defect discussions. It provides RESTful endpoints for managing chat sessions, participants, and messages. Additionally, database changes, controller classes, and unit tests support the implementation. This functionality is essential for facilitating communication on tasks within the project.

---

##### **API Enhancements**
- Updated `TaskEndpoint` and `DefectEndpoint` with chat functionality, including `getChatSessionResource()` methods.
- Integrated `ChatEndpoint` into `TaskResource` and `DefectResource`, making it possible to manage chat sessions directly on tasks and defects.

##### **Chat Implementation**
- Created `ChatEndpoint` interface defining RESTful endpoints for chat sessions, participants, and messages.
- Implemented `ChatSessionController` and `ChatMessageController` for handling chat business logic.
- Developed entities (`ChatSessionEntity`, `ChatMessageEntity`) and repositories (`ChatSessionRepository`, `ChatMessageRepository`) to manage database operations.

##### **Database Migrations**
- Added a Liquibase changelog (`liquibase-changelog-0_1_1.xml`) for database schema updates:
  - `CHAT_SESSION` table: Stores chat session details.
  - `CHAT_SESSION_PARTICIPANT` table: Manages participants in each session.
  - `CHAT_MESSAGE` table: Stores messages exchanged in chat sessions.

##### **Dependencies**
- Updated `pom.xml` to include:
  - `quarkus-resteasy`: For RESTful service functionality.
  - `quarkus-resteasy-multipart`: Adds support for handling multipart requests (for image upload).

---

##### **Chat Sessions**
- **Endpoints**: Endpoints created to allow the creation, retrieval, updating, and deletion of chat sessions associated with tasks or defects, as well as endpoints to add users as `HANDLER` or `OBSERVER`. 
- **Participant Management**: Participants in each chat session are assigned specific roles (`INITIATOR`, `HANDLER`, `OBSERVER`). Restrictions are enforced, such as allowing only one `INITIATOR` (assigned by default when calling /chat endpoint and creating a chat session).
- **Session Status**: Each chat session can be in an `OPEN`, `CLOSED`, or `ARCHIVED` status, affecting the ability to send messages.

##### **Messaging**
- **Text Messages**: Messages within a session are managed with endpoints for sending, retrieving, updating, and deleting.
- **Image Messages**: Preliminary endpoints are defined for image upload/download functionality, but ***full implementation is pending***.
- **Export Chat Logs**: The chat log of a session can be exported as JSON.

##### **Database Tables**
- `CHAT_SESSION`: Stores information about each chat session.
- `CHAT_SESSION_PARTICIPANT`: Associates participants with chat sessions and assigns roles.
- `CHAT_MESSAGE`: Stores each message, linking it to its sender and session.

---

#### **Current State**

The current implementation provides a nearly complete backend for chat functionality:

1. **API Endpoints**: Core chat endpoints are implemented, including:
   - Chat session management (create, retrieve, update status, delete, join, handle).
   - Participant management (add, retrieve, change role, remove).
   - Message handling (send, retrieve, update, delete).
   - **Placeholder endpoints for image upload/download.**

2. **Database Setup**: Liquibase migration files ensure the necessary tables are created.

3. **Testing**: 
   - Unit tests are added for `ChatSessionRepository` and `ChatMessageRepository`.
   - `ChatResourceTest` is partially implemented, with To-dos for image functionality.
---

#### **Next Steps**

1. **Complete Image Upload/Download**:
   - Implement image storage solution (e.g., cloud storage, database storage).
   - Complete the corresponding endpoints in `ChatEndpoint` and `ChatMessageController`.

2. **Refine Testing**:
   - Complete the `ChatResourceTest`.
   - Expand test coverage and validate edge cases.

3. **Clean Code**: 
   - Review and improve code organization, ensuring clean separation of concerns between controllers, repositories, and resources.
   - Implement best practices for error handling and validation.
---

#### **Unresolved Questions**

1. **Image Storage**: Which storage solution is optimal for image files?
2. **Backend-Frontend Compatibility**: Are the endpoints well-aligned with anticipated frontend requirements?

---

#### **Endpoints Summary**

| HTTP Method | Endpoint                                       | Summary                                    |
|-------------|------------------------------------------------|--------------------------------------------|
| POST        | `/chat`                                        | Create a new chat session on task          |
| GET         | `/chat/{sessionId}`                            | Get chat session details                   |
| DELETE      | `/chat/{sessionId}`                            | Delete chat session                        |
| PUT         | `/chat/{sessionId}/status`                     | Update chat session status                 |
| POST        | `/chat/{sessionId}/join`                       | Join chat session as an observer           |
| POST        | `/chat/{sessionId}/handle`                     | Join chat session as a handler             |
| POST        | `/chat/{sessionId}/participants`               | Add participants to chat session           |
| GET         | `/chat/{sessionId}/participants`               | Get participants in chat session           |
| GET         | `/chat/{sessionId}/participants/{participantId}` | Get participant details in chat session |
| PUT         | `/chat/{sessionId}/participants/{participantId}` | Change participant role in chat session |
| DELETE      | `/chat/{sessionId}/participants/{participantId}` | Remove participant from chat session    |
| GET         | `/chat/{sessionId}/participants/{participantId}/role` | Get participant role in chat session |
| POST        | `/chat/{sessionId}/messages`                   | Send a chat message in chat session        |
| GET         | `/chat/{sessionId}/messages/{messageId}`       | Get a chat message in chat session         |
| PUT         | `/chat/{sessionId}/messages/{messageId}`       | Update a chat message in chat session      |
| DELETE      | `/chat/{sessionId}/messages/{messageId}`       | Delete a chat message in chat session      |
| GET         | `/chat/{sessionId}/messages`                   | Get chat messages in chat session          |
| *TODO*: POST        | `/chat/{sessionId}/messages/upload`            | *Upload an image as a chat message*         |
| *TODO*: GET         | `/chat/{sessionId}/messages/{messageId}/download` | Download an image from a chat message  |

*Note: Image upload/download endpoints are currently placeholders and will be fully implemented after a storage solution is selected.*

---

### CODE REVIEW

File: remsfal-core/pom.xml [hunk 1]
```diff
@@ -55,6 +55,16 @@
             <artifactId>microprofile-openapi-api</artifactId>
             <version>${version.open-api}</version>
         </dependency>
+        <dependency>
```
Comment (astanik) [inline remsfal-core/pom.xml:58]: Don't use Quarkus dependencies in the core module. Only Jakarta and other specifications are allowed.

File: remsfal-core/pom.xml [hunk 2]
```diff
@@ -55,6 +55,16 @@
...
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy</artifactId>
+            <version>3.13.3</version>
+        </dependency>
+        <dependency>
```
Comment (astanik) [inline remsfal-core/pom.xml:63]: Move this dependency to /remsfal-service/pom.xml

If you need annotations like @PartType or @MultipartForm, please use the following dependency..

https://mvnrepository.com/artifact/org.jboss.resteasy/resteasy-multipart-provider/6.2.10.Final


<dependency>
    <groupId>org.jboss.resteasy</groupId>
    <artifactId>resteasy-multipart-provider</artifactId>
    <version>6.2.10.Final</version>
</dependency>

File: remsfal-core/src/main/java/de/remsfal/core/api/project/ChatEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    @Path("/{sessionId}/join")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Operation(summary = "Join chat session as a participant - observer")
+    @APIResponse(responseCode = "200", description = "Chat session joined")
+    @APIResponse(responseCode = "400", description = "Invalid input")
+    @APIResponse(responseCode = "404", description = "Project, task, or chat session not found")
+    @APIResponse(responseCode = "500", description = "Internal server error")
+    @APIResponse(responseCode = "401", description = "No user authentication provided via session cookie")
+    Response joinChatSession(
```
Comment (astanik) [inline remsfal-core/src/main/java/de/remsfal/core/api/project/ChatEndpoint.java:83]: What is the purpose of this operation?

That anyone can join a chat session without permissions? This could be very risky and potentially a security problem.

File: remsfal-core/src/main/java/de/remsfal/core/api/project/ChatEndpoint.java [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+    @Path("/{sessionId}/handle")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Operation(summary = "Join chat session as a participant - handler")
+    @APIResponse(responseCode = "200", description = "Chat session joined")
+    @APIResponse(responseCode = "400", description = "Invalid input")
+    @APIResponse(responseCode = "404", description = "Project, task, or chat session not found")
+    @APIResponse(responseCode = "500", description = "Internal server error")
+    @APIResponse(responseCode = "401", description = "No user authentication provided via session cookie")
+    Response handleChatSession(
```
Comment (astanik) [inline remsfal-core/src/main/java/de/remsfal/core/api/project/ChatEndpoint.java:96]: Same security question?

File: remsfal-service/src/main/java/de/remsfal/service/boundary/project/ChatResource.java [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+@RequestScoped
+public class ChatResource extends ProjectSubResource implements ChatEndpoint {
+
+    @Inject
+    ChatSessionController chatSessionController;
+
+    @Inject
+    ChatMessageController chatMessageController;
+
+    @Inject
```
Comment (astanik) [inline remsfal-service/src/main/java/de/remsfal/service/boundary/project/ChatResource.java:30]: Please remove this injection and use the principal from ProjectSubResource.

File: remsfal-service/src/main/java/de/remsfal/service/boundary/project/ChatResource.java [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+
+
+    @Inject
+    Logger logger;
+
+
+    @Override
+    public Response createChatSession() {
+        String userId = principal.getId(); // Get user ID from session
+        if (userId == null || userId.isBlank()) {
```
Comment (astanik) [inline remsfal-service/src/main/java/de/remsfal/service/boundary/project/ChatResource.java:41]: You do not need to check the authentication, because the authentication should be ensured by the PrincipalRequestFilter. Much more important is the authorization, because only owner and manager of a project should be able to create a new chat session. See checkPrivileges() in ProjectSubResource.

File: remsfal-service/src/main/java/de/remsfal/service/boundary/project/ChatResource.java [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+                    .entity("User is not authenticated")
+                    .build();
+        }
+
+        String projectId = uri.getPathParameters().getFirst("projectId");
+        String taskId = uri.getPathParameters().getFirst("taskId");
+        String defectId = uri.getPathParameters().getFirst("defectId");
+
+        if ((taskId == null || taskId.isBlank()) && (defectId == null || defectId.isBlank())) {
+            return Response.status(Response.Status.BAD_REQUEST)
```
Comment (astanik) [inline remsfal-service/src/main/java/de/remsfal/service/boundary/project/ChatResource.java:52]: Don't use the happy path to return an error, because you will not trigger a transaction rollback on that way. You should throw a BadRequestException()

File: remsfal-service/src/main/java/de/remsfal/service/entity/dao/ChatMessageRepository.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+    public ChatMessageEntity findChatMessageById(String messageId) {
+        return find("id = :id", Parameters.with("id", messageId))
+                .firstResultOptional()
+                .orElseThrow(() -> new NoSuchElementException("ChatMessage with ID " + messageId + " not found"));
+    }
+
+    @Transactional
+    public ChatMessageEntity sendChatMessage(String sessionId, String senderId, ContentType contentType, String content) {
+        if (sessionId == null || sessionId.isBlank()) {
```
Comment (astanik) [inline remsfal-service/src/main/java/de/remsfal/service/entity/dao/ChatMessageRepository.java:29]: This checks could be validated by BeanValidation in the boundy or on the entity.

**Review by astanik — COMMENTED** [review]
This pull request is already very huge! We should find a way to keep PRs as small as possible. I would suggest to open a new issue for the file upload support and to keep it out of this PR.

Another problem ist the missing sonar report. I'm not sure why no new lines are detected.

File: remsfal-service/src/main/java/de/remsfal/service/boundary/project/ChatResource.java (hunk 6 above)
Comment (parhamrahmani) [reply remsfal-service/src/main/java/de/remsfal/service/boundary/project/ChatResource.java:41]: I have a question regarding the authorization process. Shouldn't the chat function, *similar to a ticket*, be initiated by the client or tenant? Additionally, shouldn't the chat session, like a ticket, be managed and only closed or archived by the owner or manager? That was my initial understanding of a chat session—as something similar to a ticket

File: remsfal-core/src/main/java/de/remsfal/core/api/project/ChatEndpoint.java (hunk 3 above)
Comment (parhamrahmani) [reply remsfal-core/src/main/java/de/remsfal/core/api/project/ChatEndpoint.java:83]: My first thoughts were that the current solution allows the tenant or user to create a chat session and assigns them the role of `INITIATOR` *permanently*. Once a session is opened, a manager or owner can join the chat session as an `OBSERVER` by using the `/join` endpoint. To assign a member to handle a chat session, they can be given the role `HANDLER`. Since I assumed endpoint access required authentication, I didn’t initially anticipate any security issues.

I will add the following items to the list of future issues to be addressed:

- Authorization logic for handling, observing, and initiating sessions.
- Resolving security concerns, such as defining permissions.
- Developing a workflow to automatically assign a project member to handle a session.
- Restricting session initiation to the tenant associated with a project. I’m not sure how to implement this yet, but I’ll keep it in mind.

File: remsfal-core/src/main/java/de/remsfal/core/api/project/ChatEndpoint.java (hunk 4 above)
Comment (parhamrahmani) [reply remsfal-core/src/main/java/de/remsfal/core/api/project/ChatEndpoint.java:96]: explained on this [response](https://github.com/remsfal/remsfal-backend/pull/239#discussion_r1830916029)

**Comment by parhamrahmani** [comment]
All fixes and changes have been done. There are some major changes on endpoints and chat resource and all functions for text web chat is ready and in ChatResourceTest.java, ChatSessionRepositoryTest.java and ChatMessageRepositoryTest.java tested. 

@astanik the PR is ready for review now.

**Review by astanik — APPROVED** [review]
Very good job! :thumbsup:
